src/hotspot/share/interpreter/zero/bytecodeInterpreter.cpp

```diff
@@ -102,11 +102,16 @@
   There really shouldn't be any handles remaining to trash but this is cheap
   in relation to a safepoint.
 */
-#define SAFEPOINT                                                                 \
-    {                                                                             \
-       /* zap freed handles rather than GC'ing them */                            \
-       HandleMarkCleaner __hmc(THREAD);                                           \
-       CALL_VM(SafepointMechanism::process_if_requested(THREAD), handle_exception); \
+#define SAFEPOINT                                                    \
+    {                                                                \
+       /* zap freed handles rather than GC'ing them */               \
+       HandleMarkCleaner __hmc(THREAD);                              \
+       if (SafepointMechanism::should_process(THREAD)) {             \
+         CALL_VM(SafepointMechanism::process_if_requested(THREAD);   \
+                 if (THREAD->has_special_runtime_exit_condition()) { \
+                   THREAD->handle_special_runtime_exit_condition();  \
```

Comment:
Please add explicit arg: "(true /* check asyncs */)"

Using a multi-statement code block for the CALL_VM macro looks odd to me, but I find these zero macros unpleasant to begin with.

Reply:
Yes, that's also why I liked better the TIVFJ wrapper. But if we can agree on the helper method then that would change to a single line.